src/extensions/rppgrpc/rppgrpc/client_reactor.hpp

```diff
@@ -0,0 +1,303 @@
+//                  ReactivePlusPlus library
+//
+//          Copyright Aleksey Loginov 2023 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/observables/fwd.hpp>
+
+#include <rpp/subjects/publish_subject.hpp>
+
+#include <grpcpp/support/client_callback.h>
+#include <rppgrpc/fwd.hpp>
+#include <rppgrpc/utils/exceptions.hpp>
+
+#include <deque>
+
+namespace rppgrpc::details
+{
+    template<rpp::constraint::decayed_type Request, rpp::constraint::observer Observer>
+    class client_bidi_reactor final : public grpc::ClientBidiReactor<Request, rpp::utils::extract_observer_type_t<Observer>>
+    {
+        using Response = rpp::utils::extract_observer_type_t<Observer>;
+        using Base     = grpc::ClientBidiReactor<Request, Response>;
+
+    public:
+        template<rpp::constraint::observable_of_type<Request> Observable, rpp::constraint::decayed_same_as<Observer> TObserver>
+        client_bidi_reactor(const Observable& messages, TObserver&& events)
+            : m_observer{std::forward<TObserver>(events)}
+            , m_disposable{messages.subscribe_with_disposable([this]<rpp::constraint::decayed_same_as<Request> T>(T&& message) {
+                std::lock_guard lock{m_write_mutex};
+                m_write.push_back(std::forward<T>(message));
+                if (m_write.size() == 1)
+                    Base::StartWrite(&m_write.front()); },
```

Comment:
Can zip operator be used to avoid using locks? One stream can emit write ready messages and other emit proto messages

Reply:
Not sure if I get your point. Anyway, zip operator uses mutex itself :)